src/session.rs

```diff
@@ -263,31 +270,48 @@ impl Session {
                 )
             })?;
 
-            inner.rc(handshake(inner.raw, stream))
+            inner.rc(handshake(inner.raw, stream.as_ref()))
         }
     }
 
-    /// The session takes ownership of the socket provided.
-    /// You may use the tcp_stream() method to obtain a reference
-    /// to it later.
+    /// The session takes ownership of the stream provided.
+    /// You may use the tcp_stream() method to obtain the raw fd later.
+    ///
+    /// It is also highly recommended that the stream provided is not used
+    /// concurrently elsewhere for the duration of this session as it may
+    /// interfere with the protocol.
+    #[cfg(unix)]
+    pub fn set_tcp_stream<S: 'static + AsRawFd>(&mut self, stream: S) {
+        let mut inner = self.inner();
+        let _ = inner.tcp.replace(Box::new(stream));
+    }
+
+    /// The session takes ownership of the stream provided.
+    /// You may use the tcp_stream() method to obtain the raw socket later.
     ///
     /// It is also highly recommended that the stream provided is not used
     /// concurrently elsewhere for the duration of this session as it may
     /// interfere with the protocol.
-    pub fn set_tcp_stream(&mut self, stream: TcpStream) {
+    #[cfg(windows)]
+    pub fn set_tcp_stream<S: 'static + AsRawSocket>(&mut self, stream: S) {
         let mut inner = self.inner();
-        let _ = inner.tcp.replace(stream);
+        let _ = inner.tcp.replace(Box::new(stream));
     }
 
-    /// Returns a reference to the stream that was associated with the Session
-    /// by the Session::handshake method.
-    pub fn tcp_stream(&self) -> Option<MappedMutexGuard<TcpStream>> {
+    /// Returns the raw fd of the stream that was associated with the Session by
+    /// the Session::handshake method.
+    #[cfg(unix)]
+    pub fn tcp_stream(&self) -> Option<RawFd> {
```

Comment:
I know that I suggested this, but I wonder what you think about implementing `AsRawFd` and `AsRawSocket` on `Session` in the place of these methods?  Those traits don't allow for the stream to be `None`, but it's probably fine to have those implementations panic if used without assigning a stream.   I don't think anyone should really be using this method or those traits anyway.

Reply:
I think I like the panic approach with implementing those traits more. I'll change that.